torch/tensortypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,215 @@
+        ret = ret.to(device)
+    return ret
+
+class NestedTensor():
+    # The attributes must match across all constiuents
+    # and default to the empty Tensor's if the given list
+    # is empty.
+    #
+    # The NestedTensor's attributes then become that of its
+    # constiuents.
+    #
+    # Attributes:
+    #     dim
+    #     layout
+    #     device
+    #     dtype
+    #     requires_grad
+    #     is_pinned
+    def __init__(self, tensors):
+        for tensor in tensors:
+            assert torch.is_tensor(tensor)
+        self.tensors = tensors
+        if len(tensors):
+            self.dim = tensors[0].dim()
+            self.layout = tensors[0].layout
+            self.device = tensors[0].device
+            self.dtype = tensors[0].dtype
+            requires_grad = tensors[0].requires_grad
+            is_pinned = tensors[0].is_pinned()
+            for tensor in tensors:
+                if not (self.dim == tensor.dim() and
+                        self.layout == tensor.layout and
+                        self.device == tensor.device and
+                        self.dtype == tensor.dtype and
+                        requires_grad == tensor.requires_grad and
+                        is_pinned == tensor.is_pinned()):
+                    raise ValueError("Each passed Tensor "
+                                     "must match in dim, layout, "
+                                     "device, dtype and requires_grad")
+        else:
+            # Carrying around information as member variables vs.
+            # checking one entry of the owned Tensors is annoying
+            # and error-prone. Carrying around an is_empty attribute
+            # to hide the fact that we carry around a list with a
+            # single empty Tensor is also annoying and error-prone.
+            # Both are not worth it for a minor feature.
+            raise ValueError("We do not support empty lists for now.")
+
+    def __getattribute__(self, attr):
+        if attr == 'shape':
+            raise NotImplementedError()
+        if attr == 'requires_grad':
+            return self.tensors[0].requires_grad
+        if attr == 'grad_fn':
+            # We're assuming len(self.tensors) is non-zero
+            if self.tensors[0].grad_fn is None:
+                for i in range(1, len(self.tensors)):
+                    assert self.tensors[i].grad_fn is None
+                return None
+
+            def create_grad_fn(nested_tensor):
+                def _func(*args, **kwargs):
+                    for tensor in nested_tensor.tensors:
+                        tensor.grad_fn(*args, **kwargs)
+                return _func
+            return create_grad_fn(self)
+        if attr == 'grad':
+            grads = []
+            for tensor in self.tensors:
+                grads.append(tensor.grad)
+            if all(grad is None for grad in grads):
+                return None
+            else:
+                return NestedTensor(grads)
```

Comment:
This NestedTensor had better alias the original grads, because I am definitely going to want `nested_tensor.grad.zero_()` to work.

Reply:
Yes, the NestedTensor constructor simply takes the list and assigns it to self.tensors. Then this should never be a problem, right?

Reply:
I think what's confusing is that `torch.nested_tensor` does a copy, but the direct `NestedTensor` constructor does not. We probably should discourage people from using the constructor directly, it's not in line with the current conventions on Tensor.

Reply:
Yes, this function is internal and as such its ok to call the NestedTensor constructor, but in general it should not be called by users just as with torch.Tensor. We have the as_tensor constructor for non-copy construction and in line with that we have the as_nestedtensor function in this file.